core/src/main/java/io/ddf/DDFCache.java

```diff
@@ -2,52 +2,166 @@
+
+    mLog.info(String.format("Maximum number of ddfs in cache %s", maxNumberOfDDFs));
+    mLog.info(String.format("DDF's time to idle = %s seconds", timeToIdle));
+
+    mDDFCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFs).expireAfterAccess(timeToIdle, TimeUnit.SECONDS).
+        recordStats().removalListener(new DDFRemovalListener())
+        .build();
+
+    // Don't do time based eviction for DDF with name
+    mDDFwithNameCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFsWName).recordStats().removalListener(new DDFRemovalListener())
+        .build();
+  }
+
+  private boolean containsDDF(UUID uuid) {
+    return (mDDFCache.asMap().containsKey(uuid) || mDDFwithNameCache.asMap().containsKey(uuid));
+  }
 
-  private Map<UUID, DDF> mDDFs = new ConcurrentHashMap<UUID, DDF>();
   private Map<String, UUID> mUris = new ConcurrentHashMap<String, UUID>();
 
   public void addDDF(DDF ddf) throws DDFException {
-    mDDFs.put(ddf.getUUID(), ddf);
+    if(ddf.getUUID() == null) {
+      throw new DDFException("uuid is null");
+    }
+    Lock lock = uuidLocks.get(ddf.getUUID());
+    try {
+      if(lock.tryLock(120, TimeUnit.SECONDS)) {
+        try {
+          if (ddf.getName() == null) {
+            mDDFCache.put(ddf.getUUID(), ddf);
+          } else {
+            mDDFwithNameCache.put(ddf.getUUID(), ddf);
+          }
+        } finally {
+          lock.unlock();
+        }
+      } else {
+        throw new DDFException("Timeout getting DDF");
+      }
+    } catch (InterruptedException e) {
+
+    }
+  }
+
+  public CacheStats getCacheStats() {
+    return mDDFCache.stats().plus(mDDFwithNameCache.stats());
   }
 
-  public void removeDDF(DDF ddf) throws DDFException {
-    mDDFs.remove(ddf.getUUID());
+  public synchronized void removeDDF(DDF ddf) throws DDFException {
+    mDDFCache.invalidate(ddf.getUUID());
+    mDDFwithNameCache.invalidate(ddf.getUUID());
     if(ddf.getUri() != null) {
       mUris.remove(ddf.getUri());
     }
   }
 
   public DDF[] listDDFs() {
-    return this.mDDFs.values().toArray(new DDF[] {});
+    List<DDF> ddfs = new ArrayList<DDF>();
+    ddfs.addAll(mDDFCache.asMap().values());
+    ddfs.addAll(mDDFwithNameCache.asMap().values());
+    return ddfs.toArray(new DDF[]{});
   }
 
-  public DDF getDDF(UUID uuid) throws DDFException {
-    DDF ddf = mDDFs.get(uuid);
-    if(ddf == null) {
-      throw new DDFException(String.format("Cannot find ddf with uuid %s", uuid));
-
-    } else {
-      return ddf;
-    }
+  public long size() {
+    return mDDFCache.size() + mDDFwithNameCache.size();
   }
 
-  public boolean hasDDF(UUID uuid) {
-    DDF ddf = mDDFs.get(uuid);
-    return ddf != null;
+  /**
+   *  Try to get DDF from both mDDFCache and mDDFWithNameCache
+   *  if DDF is not present in both, do a restore and put it into the respective cache
+   * @param uuid
+   * @return
+   * @throws DDFException
+   */
+  public DDF getDDF(UUID uuid) throws DDFException {
```

Comment:
Agree, if you couldn't use CacheLoader, you will have to provide an other load function to update your two cache. That function is threadsafe, not locking on getDDF function like that.
let say:
```
getDDF(UUID id)
 ddf = priorityCache.getIfPresent( id )
 if( ddf is null)
   ddf = normalCache.getIfPresent( id)
if (ddf is still null)
   ddf = _custom_cache_loader( id )
if(ddf is still null) throw new NotFoundException( ...)
return ddf
```